### ISSUE TITLE
Documentation problem['groups']

### ISSUE THREAD

**Issue by rJonatan** [SALib/SALib#161]
Hi!
I'd like to understand better the meaning of 'groups' in the problem definition. Could someone please explain how to use it?

Currently, the API documentation only comments groups for the method of Morris (http://salib.readthedocs.io/en/latest/api.html#method-of-morris), but judging from the source code it applies also to Sobol: http://salib.readthedocs.io/en/latest/_modules/SALib/analyze/sobol.html. Not for FAST however.

**Comment by willu47** [comment]
Hi @rJonatan, thanks a lot for the issue.  You are absolutely right, the documentation on groups is poor.  I implemented the groups functionality for Morris.  You can read more about the approach with groups in section 3.5 of [Saltelli 2008](http://www.andreasaltelli.eu/file/repository/A_Saltelli_Marco_Ratto_Terry_Andres_Francesca_Campolongo_Jessica_Cariboni_Debora_Gatelli_Michaela_Saisana_Stefano_Tarantola_Global_Sensitivity_Analysis_The_Primer_Wiley_Interscience_2008_.pdf).

In SALib, to define a problem file with groups, you add an extra column in which you name the group to which a parameter belongs. For example:
```
x1 -3.14 3.14 Group_1
x2 -3.14 3.14 Group_2
x3 -3.14 3.14 Group_2
```
This is loaded into SALib and creates a problem dictionary with three parameters with ranges [0, 1] and two groups.  `Group 1` contains `Param 1` and `Param 2` and `Group 2` contains `Param 3`.  The dict which is populated in SALib (and which you can pass directly to the sample and analyze methods) looks like this:
```python
{
'groups': ([0, 1, 1], ['Group_1', 'Group_2']),
'names': ['x1', 'x2', 'x3'],
'num_vars': 3,
'bounds': [[-3.14, 3.14], [-3.14, 3.14], [-3.14, 3.14]]
}
```
The `groups` value contains a tuple in which the first entry is a list which maps the parameters to groups, and the second entry contains a list of the group names. That is all there is to it.

@jdherman - I dimly remember that groups were added to Sobol, could you confirm?

**Comment by jdherman** [comment]
Good find. Yes, the groups feature is added to Sobol, but not the other methods. Unfortunately I'm seeing that it's using a slightly different API than what Will mentioned above.

Here is a working Sobol example with groups:
```python
problem = {
  'groups': ['Group_1', 'Group_2', 'Group_2'],
  'names': ['x1', 'x2', 'x3'],
  'num_vars': 3,
  'bounds': [[-3.14, 3.14], [-3.14, 3.14], [-3.14, 3.14]]
}

# Generate samples
param_values = saltelli.sample(problem, 1000)

# Run model (example)
Y = Ishigami.evaluate(param_values)

# Perform analysis
Si = sobol.analyze(problem, Y, print_to_console=True)
```

The difference is the `groups` value in the problem dictionary. It looks like I'll have to change how the Sobol method interprets this to match Will's version, so it's not only a question of documentation. I will try to get to this next week. In the meantime @rJonatan, Sobol with groups should work as above in v1.10.

**Comment by rJonatan** [comment]
Thanks for the explanations, @willu47  and @jdherman . So, do I understand correctly that the purpose of the groups syntax is to be able to sample groups of variation factors together? And consequently we then get results of the sensitivity analysis for each group rather than for each individual factor?

Do you plan to implement this feature also for other methods - such as FAST - or is there a reason you've implemented it for Morris and Sobol in particular?

**Comment by willu47** [comment]
@jdherman let me double check the morris groups API...  I could be mistaken.

@rJonatan - yes, that's the purpose. You thereby reduce the dimensionality, shrink the number of samples required, reducing the computational burden of conducting the analysis.  The cost is less explanatory power, but the benefit is that the approach is still global - all parameters are moved over their ranges.

**Comment by willu47** [comment]
@jdherman - see [line 144](https://github.com/SALib/SALib/blob/f44161489c64db5c816dc081ffb8d1d532f62c4d/SALib/util/__init__.py#L144) for the groups implementation returned from the problem file reader.

**Comment by jdherman** [comment]
Ah ok I think I see the problem. The docstrings say something different from what the function actually does. `groups` should be a unique list of strings (group names), everywhere throughout the library. Both Morris and Sobol are set up to use this.

When the groups matrix needs to be calculated, we have the separate `compute_groups_matrix` function in `util/__init__.py`. 

I will go through and fix the docstrings. PR in a minute.

**Comment by jdherman** [comment]
Fixed by #163
